# NumPy and SciPy for finite difference methods

Computational Finance with Python

[Alet Roux](https://www.york.ac.uk/maths/staff/alet-roux/) ([Department
of Mathematics](https://maths.york.ac.uk), University of York)

Click on the following to open this file in Google Colab:

<figure>
<a
href="https://colab.research.google.com/github/aletroux/comp-finance-python/blob/main/demonstrations/D04_NumPy_SciPy_finite_differences_slides.ipynb"><img
src="https://colab.research.google.com/assets/colab-badge.svg"
alt="Open In Colab" /></a>
<figcaption>Open In Colab</figcaption>
</figure>

# Diagonal matrices and eigenvalues with NumPy

## Diagonal matrices

### Identity matrix

-   The square identity matrix can be created in a number of ways.
-   Using `numpy.eye`:

In [1]:
import numpy as np
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

-   Using `numpy.identity`:

In [2]:
np.identity(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

------------------------------------------------------------------------

### Creating diagonal matrices

-   The function `numpy.diag` can be used to create diagonal matrices.
-   Usage:

``` python
np.diag(<array with diagonal entries>, <diagonal index = 0>)
```

-   The diagonal index is:
    -   0 for the main diagonal.
    -   $k>0$, for the $k^\text{th}$ diagonal *above* the main diagonal.
    -   $-k<0$, for the $k^\text{th}$ diagonal *below* the main
        diagonal.

------------------------------------------------------------------------

### Examples

-   Create a $4\times 4$ matrix with 1 on the main diagonal, 0.5 on the
    diagonal above it and -0.5 on the diagonal below it.

In [3]:
d = np.ones(3)
A = np.eye(4) + np.diag(0.5*d,1) - np.diag(0.5*d,-1)
print("A =\n", A)

A =
 [[ 1.   0.5  0.   0. ]
 [-0.5  1.   0.5  0. ]
 [ 0.  -0.5  1.   0.5]
 [ 0.   0.  -0.5  1. ]]

-   Create a $5\times 5$ matrix with 1, …, 5 on the main diagonal, and
    3, 2, 1 on the second diagonal below it.

In [4]:
d = np.arange(1, 6)
l = np.arange(3, 0, -1)
np.diag(d) + np.diag(l,-2)

array([[1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [3, 0, 3, 0, 0],
       [0, 2, 0, 4, 0],
       [0, 0, 1, 0, 5]])

------------------------------------------------------------------------

### Extracting diagonals from matrices

-   The function `numpy.diag` can also be used to extract diagonals from
    matrices.
-   Usage:

``` python
np.diag(<2-dimensional square array>, <diagonal index = 0>)
```

-   Example: Extract the main diagonal from the matrix `A` of the
    previous slide.

In [5]:
np.diag(A, 0)

array([1., 1., 1., 1.])

In [6]:
np.diag(A)

array([1., 1., 1., 1.])

-   Example: Extract the diagonal above the main diagonal.

In [7]:
np.diag(A, 1)

array([0.5, 0.5, 0.5])

-   Example: Extract the diagonal below the main diagonal.

In [8]:
np.diag(A, -1)

array([-0.5, -0.5, -0.5])

## Eigenvalues and eigenvectors

-   The Linear Algebra module of NumPy (NumPy Developers (2024)) can be
    used to find the eigenvalues and eigenvectors of a matrix.

In [9]:
import numpy.linalg as npla

-   Usage for general (square) matrices:

``` python
[eigenvalues, eigenvectors] = npla.eig(<2-dimensional square array>)
```

-   To obtain only eigenvalues:

``` python
eigenvalues = npla.eigvals(<2-dimensional square array>)
```

-   Usage for symmetric (Hermitian) matrices:

``` python
[eigenvalues, eigenvectors] = npla.eigh(<2-dimensional symmetric (Hermitian) array>)
eigenvalues = npla.eigvalsh(<2-dimensional symmetric (Hermitian) array>)
```

-   Use the appropriate routines for symmetric matrices for a
    performance boost.

------------------------------------------------------------------------

### Example: General matrix

In [10]:
print("A =\n", A)

A =
 [[ 1.   0.5  0.   0. ]
 [-0.5  1.   0.5  0. ]
 [ 0.  -0.5  1.   0.5]
 [ 0.   0.  -0.5  1. ]]

-   Eigenvalues can be complex-valued (`j` represents the imaginary
    number):

In [11]:
[eigenvalues, eigenvectors] = npla.eig(A)
print("Eigenvalues:\n", eigenvalues)
print("Eigenvectors:\n", eigenvectors)

Eigenvalues:
 [1.+0.80901699j 1.-0.80901699j 1.+0.30901699j 1.-0.30901699j]
Eigenvectors:
 [[-3.71748034e-01+1.66533454e-16j -3.71748034e-01-1.66533454e-16j
   6.01500955e-01+0.00000000e+00j  6.01500955e-01-0.00000000e+00j]
 [ 5.55111512e-16-6.01500955e-01j  5.55111512e-16+6.01500955e-01j
  -6.93889390e-16+3.71748034e-01j -6.93889390e-16-3.71748034e-01j]
 [ 6.01500955e-01+0.00000000e+00j  6.01500955e-01-0.00000000e+00j
   3.71748034e-01-5.55111512e-17j  3.71748034e-01+5.55111512e-17j]
 [-1.66533454e-16+3.71748034e-01j -1.66533454e-16-3.71748034e-01j
  -1.11022302e-15+6.01500955e-01j -1.11022302e-15-6.01500955e-01j]]

-   Calculating absolute values of eigenvalues:

In [12]:
np.abs(npla.eigvals(A))

array([1.28627699, 1.28627699, 1.0466573 , 1.0466573 ])

------------------------------------------------------------------------

### Example: Symmetric matrix

-   Create a symmetric matrix:

In [13]:
B = abs(A)
print("B =\n", B)

B =
 [[1.  0.5 0.  0. ]
 [0.5 1.  0.5 0. ]
 [0.  0.5 1.  0.5]
 [0.  0.  0.5 1. ]]

-   The eigenvalues of symmetric matrices are real-valued:

In [14]:
npla.eigvalsh(B)

array([0.19098301, 0.69098301, 1.30901699, 1.80901699])

In [15]:
npla.eigvals(B)

array([1.80901699, 1.30901699, 0.19098301, 0.69098301])

-   Taking advantage of symmetry gives a performance boost:

In [16]:
%timeit npla.eigvalsh(B)

5.26 μs ± 67.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

In [17]:
%timeit npla.eigvals(B)

11.7 μs ± 274 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

# Sparse matrices with SciPy

-   A matrix is called **sparse** if it contains many zero entries *and
    it’s worth taking advantage of*.
-   Sparse matrices are common in applications such as finite difference
    methods, where matrices are large but (typically) of a diagonal
    nature.
-   The `scipy.sparse` module (The SciPy Community (2024e)) is designed
    to work with sparse matrices.
-   To use this module in your work:

In [18]:
import scipy.sparse as ssparse

## Toy example

-   The following $6\times6$ NumPy matrix has 20 zero entries and 16
    non-zero entries:

In [19]:
d = np.ones(5)
C = np.eye(6) + 0.5*np.diag(d,1) - 0.5*np.diag(d,-1)
print("C =\n", C)

C =
 [[ 1.   0.5  0.   0.   0.   0. ]
 [-0.5  1.   0.5  0.   0.   0. ]
 [ 0.  -0.5  1.   0.5  0.   0. ]
 [ 0.   0.  -0.5  1.   0.5  0. ]
 [ 0.   0.   0.  -0.5  1.   0.5]
 [ 0.   0.   0.   0.  -0.5  1. ]]

-   It can be converted into a more efficient *DIAgonal storage*
    representation, which records only the non-zero entries (keeping
    diagonals together):

In [20]:
C_dia = ssparse.dia_array(C)
print ("Number of non-zero entries:", C_dia.size)
print ("Diagonal offsets:", C_dia.offsets)
print ("Data:\n", C_dia.data)

Number of non-zero entries: 16
Diagonal offsets: [-1  0  1]
Data:
 [[-0.5 -0.5 -0.5 -0.5 -0.5  0. ]
 [ 1.   1.   1.   1.   1.   1. ]
 [ 0.   0.5  0.5  0.5  0.5  0.5]]

------------------------------------------------------------------------

## Toy example (cont)

-   Conversion back to a NumPy array is straightforward:

In [21]:
C_dia.toarray()

array([[ 1. ,  0.5,  0. ,  0. ,  0. ,  0. ],
       [-0.5,  1. ,  0.5,  0. ,  0. ,  0. ],
       [ 0. , -0.5,  1. ,  0.5,  0. ,  0. ],
       [ 0. ,  0. , -0.5,  1. ,  0.5,  0. ],
       [ 0. ,  0. ,  0. , -0.5,  1. ,  0.5],
       [ 0. ,  0. ,  0. ,  0. , -0.5,  1. ]])

-   Sparse matrices can be treated just like ordinary NumPy matrices.
-   For example:

In [22]:
print("C**2 =\n", (C_dia @ C_dia).toarray())

C**2 =
 [[ 0.75  1.    0.25  0.    0.    0.  ]
 [-1.    0.5   1.    0.25  0.    0.  ]
 [ 0.25 -1.    0.5   1.    0.25  0.  ]
 [ 0.    0.25 -1.    0.5   1.    0.25]
 [ 0.    0.    0.25 -1.    0.5   1.  ]
 [ 0.    0.    0.    0.25 -1.    0.75]]

## Sparse diagonal matrices

-   A sparse diagonal matrix can be created using the
    `scipy.sparse.diags` function (The SciPy Community (2024b)). Usage:

``` python
<array> = ssparse.diags(<diagonals>, offsets=0, shape=None)
```

-   `offsets` is a sequence of integers (or just one integer)
    corresponding to the diagonal indices:
    -   0 for the main diagonal.
    -   $k>0$, for the $k^\text{th}$ diagonal *above* the main diagonal.
    -   $-k<0$, for the $k^\text{th}$ diagonal *below* the main
        diagonal. If omitted, assumes the main diagonal.
-   `diagonals` is a sequence of arrays containing the array diagonals
    corresponding to `offsets`. Can be replaced by a number if the
    diagonal is constant.
-   `shape` optionally gives the size $(n,n)$ of the matrix, only needed
    if it cannot be inferred from the other arguments.

------------------------------------------------------------------------

### An important note about `scipy.sparse.diags`

-   Since version 1.11 of the SciPy library, the function
    `scipy.sparse.diags` has been superceded by the new function
    `scipy.sparse.diags_array` (The SciPy Community (2024c)), which has
    the same arguments.

-   It is recommended to use the new function where it is available.
    However it does not work in version 1.11.4, and so
    `scipy.sparse.diags` should be used on Colab in 2024.

------------------------------------------------------------------------

### Example: constant diagonals

-   Let us recreate the matrix $C$ of the toy example:

In [23]:
diagonals = [0.5, 1, -0.5]
offsets = [1, 0, -1]
C = ssparse.diags(diagonals, offsets = offsets, shape = (6,6))

print ("Number of non-zero entries:", C.size)
print ("Diagonal offsets:", C.offsets)
print ("Data:\n", C.data)

Number of non-zero entries: 16
Diagonal offsets: [ 1  0 -1]
Data:
 [[ 0.   0.5  0.5  0.5  0.5  0.5]
 [ 1.   1.   1.   1.   1.   1. ]
 [-0.5 -0.5 -0.5 -0.5 -0.5  0. ]]

In [24]:
print ("NumPy array:\n", C.toarray())

NumPy array:
 [[ 1.   0.5  0.   0.   0.   0. ]
 [-0.5  1.   0.5  0.   0.   0. ]
 [ 0.  -0.5  1.   0.5  0.   0. ]
 [ 0.   0.  -0.5  1.   0.5  0. ]
 [ 0.   0.   0.  -0.5  1.   0.5]
 [ 0.   0.   0.   0.  -0.5  1. ]]

------------------------------------------------------------------------

### Example: variable diagonals

-   Create a $10\times 10$ matrix with 1, …, 10 on the main diagonal,
    and 8, …, 1 on the second diagonal below it.

In [25]:
diagonals = [np.arange(1,11), np.arange(8, 0, -1)]
offsets = [0, -2]
D = ssparse.diags(diagonals, offsets = offsets)

print ("Number of non-zero entries:", D.size)
print ("Diagonal offsets:", D.offsets)
print ("Data:\n", D.data)

Number of non-zero entries: 18
Diagonal offsets: [ 0 -2]
Data:
 [[ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
 [ 8.  7.  6.  5.  4.  3.  2.  1.  0.  0.]]

In [26]:
print ("NumPy array:\n", D.toarray())

NumPy array:
 [[ 1.  0.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  2.  0.  0.  0.  0.  0.  0.  0.  0.]
 [ 8.  0.  3.  0.  0.  0.  0.  0.  0.  0.]
 [ 0.  7.  0.  4.  0.  0.  0.  0.  0.  0.]
 [ 0.  0.  6.  0.  5.  0.  0.  0.  0.  0.]
 [ 0.  0.  0.  5.  0.  6.  0.  0.  0.  0.]
 [ 0.  0.  0.  0.  4.  0.  7.  0.  0.  0.]
 [ 0.  0.  0.  0.  0.  3.  0.  8.  0.  0.]
 [ 0.  0.  0.  0.  0.  0.  2.  0.  9.  0.]
 [ 0.  0.  0.  0.  0.  0.  0.  1.  0. 10.]]

------------------------------------------------------------------------

### Example: efficiency of sparse matrices

-   Let us compare the speed of multiplying a $500\times500$ matrix by
    itself, using both sparse arrays and standard NumPy arrays.

In [27]:
diagonals = [0.5, 1, 0.5]
offsets = [1, 0, -1]
E = ssparse.diags(diagonals, offsets = offsets, shape = (500,500))
Earray = E.toarray()
print ("Number of non-zero entries:", E.size)
print ("Number of entries in total:", 500**2)

Number of non-zero entries: 1498
Number of entries in total: 250000

-   As a sparse matrix:

In [28]:
%timeit E @ E

357 μs ± 18.9 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

-   As a NumPy array:

In [29]:
%timeit Earray @ Earray

851 μs ± 11.6 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

------------------------------------------------------------------------

### Sparse matrices in other formats

-   We focus on diagonal matrices because they are most relevant for our
    work. However, there are different representations for sparse
    matrices that work well in different contexts. The SciPy Community
    (2024e) provides detail on different representations, classes, as
    well as functions for building sparse matrices and working with
    them.

-   One particular case is *Compressed Sparse Column*, which is very
    efficient for inverting matrices and solving linear equations. The
    SciPy Community (2024a) provides full information. Convert a
    diagonal sparse matrix into this format as follows:

In [30]:
Ecsc = E.tocsc()
Ecsc

<Compressed Sparse Column sparse matrix of dtype 'float64'
    with 1498 stored elements and shape (500, 500)>

-   Convert a sparse matrix in any other format to diagonal format as
    follows:

In [31]:
Edia = Ecsc.todia()
Edia

<DIAgonal sparse matrix of dtype 'float64'
    with 1498 stored elements (3 diagonals) and shape (500, 500)>

## Linear algebra with sparse matrices

-   The `scipy.sparse.linalg` submodule provides functionality for doing
    linear algebra with sparse arrays.

-   Include this module in your code as follows:

In [32]:
import scipy.sparse.linalg as ssla

-   In the following slides we will cover the basics of solving linear
    equations, eigenvalues and eigenvectors, but this is only a very
    small part of the available functionality. Consult the documentation
    (The SciPy Community (2024d)) for further information.

-   **Warning**: do not apply NumPy routines to sparse arrays because
    NumPy might not properly convert them, leading to unexpected and
    incorrect results. Do the following:

    -   Check if SciPy has its own implementation.
    -   If not, then convert the sparse array to a NumPy array before
        applying the method.

------------------------------------------------------------------------

### Inverting a sparse matrix

-   The Compressed Sparse Column format is most efficient for computing
    inverses and solving linear equations.

In [33]:
Ecsc = E.tocsc()

-   The inverse of a matrix can be computed as follows:

In [34]:
ssla.inv(Ecsc)

<Compressed Sparse Column sparse matrix of dtype 'float64'
    with 250000 stored elements and shape (500, 500)>

-   The number of stored elements increases, even for tridiagonal
    matrices. This means that multiplying the inverse of an $n\times n$
    matrix by a vector takes $n^2$ calculations, even though the matrix
    itself may be sparse.

-   For this reason, calculating the inverse is not always the most
    efficient way of solving a linear system.

------------------------------------------------------------------------

### Solving linear systems with sparse matrices

-   Use `scipy.sparse.linalg.spsolve` to solve a system $Ex=b$ of linear
    equations:

In [35]:
b = np.ones((500,1))
x = ssla.spsolve(Ecsc, b)
print("Size of x:", x.size)
print("First 5 elements of x:", x[:5])

Size of x: 500
First 5 elements of x: [0.99800399 0.00399202 0.99401198 0.00798403 0.99001996]

-   Use `scipy.sparse.linalg.factorized` to create a function that
    solves systems of the form $Ex=b$ (where $b$ is given as a
    parameter):

In [36]:
solveit = ssla.factorized(Ecsc)
x = solveit(b)
print("Size of x:", x.size)
print("First 5 elements of x:", x[:5].T)

Size of x: 500
First 5 elements of x: [[0.99800399 0.00399202 0.99401198 0.00798403 0.99001996]]

------------------------------------------------------------------------

### Eigenvalues and eigenvectors of sparse matrices

-   Usage for general (square) sparse matrices:

``` python
[eigenvalues, eigenvectors] = ssla.eigs(<square sparse matrix>, k=6, which = 'LM', return_eigenvectors=True)
```

-   `k` is the number of eigenvalues (and optionally eigenvectors) to
    find, must be less than the dimension of the matrix. It is not
    generally possible to find all eigenvalues of sparse matrices.

-   `which` indicates which eigenvalues to find. The most relevant
    options for our work are `LM` for *largest magnitude* and `SM` for
    *smallest magnitude*.

-   Use `return_eigenvectors=False` if eigenvectors are not required.

-   Usage for symmetric (Hermitian) matrices:

``` python
[eigenvalues, eigenvectors] = ssla.eigsh(<square sparse matrix>, k=6, which = 'LM', return_eigenvectors=True)
```

-   Use the appropriate routines for symmetric matrices for a
    performance boost.

------------------------------------------------------------------------

#### Example: General matrix

-   Create a $500\times 500$ matrix with 1 on the main diagonal, 0.5 on
    the diagonal above it and -0.5 on the diagonal below it.

In [37]:
diagonals = [0.5, 1, -0.5]
offsets = [1, 0, -1]
F = ssparse.diags(diagonals, offsets = offsets, shape = (500,500))

-   Eigenvalues can be complex-valued (`j` represents the imaginary
    number).
-   Use `LM` or `SM` to find the smallest or largest eigenvalues (in
    absolute value):

In [38]:
largest_eigenvalues = ssla.eigs(F, k=4, which = 'LM', return_eigenvectors=False)
print("Four largest eigenvalues:", largest_eigenvalues)
print("In absolute value:       ", np.abs(largest_eigenvalues))

Four largest eigenvalues: [1.+0.99992136j 1.-0.99992136j 1.+0.99998034j 1.-0.99998034j]
In absolute value:        [1.41415796 1.41415796 1.41419966 1.41419966]

In [39]:
smallest_eigenvalues = ssla.eigs(F, k=4, which = 'SM', return_eigenvectors=False)
print("Four smallest eigenvalues:", smallest_eigenvalues)
print("In absolute value:        ", np.abs(smallest_eigenvalues))

Four smallest eigenvalues: [1.+0.00940583j 1.-0.00940583j 1.+0.00313532j 1.-0.00313532j]
In absolute value:         [1.00004423 1.00004423 1.00000492 1.00000492]

------------------------------------------------------------------------

#### Example: Symmetric matrix

-   Create a $500\times 500$ matrix with 1 on the main diagonal, 0.5 on
    the diagonal above it and 0.5 on the diagonal below it.

In [40]:
diagonals = [0.5, 1, 0.5]
offsets = [1, 0, -1]
E = ssparse.diags(diagonals, offsets = offsets, shape = (500,500))

-   The eigenvalues of symmetric matrices are real-valued.

In [41]:
largest_eigenvalues = ssla.eigsh(E, k=4, which = 'LM', return_eigenvectors=False)
print("Four largest eigenvalues:", largest_eigenvalues)

Four largest eigenvalues: [1.99968545 1.99982306 1.99992136 1.99998034]

In [42]:
smallest_eigenvalues = ssla.eigsh(E, k=4, which = 'SM', return_eigenvectors=False)
print("Four smallest eigenvalues:", smallest_eigenvalues)

Four smallest eigenvalues: [3.14551320e-04 1.76939176e-04 7.86409221e-05 1.96604238e-05]

-   Taking advantage of symmetry gives a small performance boost:

In [43]:
%timeit ssla.eigs(E, k=4, which = 'LM', return_eigenvectors=False)

104 ms ± 2.95 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [44]:
%timeit ssla.eigsh(E, k=4, which = 'LM', return_eigenvectors=False)

88 ms ± 2.91 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# Tridiagonal matrix solver

-   Recall the algorithm for solving $$Ax = b,$$ where $A$ is the
    tridiagonal matrix $$A = \begin{bmatrix}
    d_1 & u_1 & 0 & \cdots & 0 \\
    l_2 & d_2 & u_2 & \ddots & \vdots \\
    0 & \ddots & \ddots & \ddots & 0 \\
    \vdots & \ddots & l_{n-1} & d_{n-1} & u_{n-1} \\
    0 & \cdots & 0 & l_n & d_n
    \end{bmatrix}.$$

## Tridiagonal matrix solver procedure

1.  Set $$\begin{aligned}
    d'_1 &= d_1, & b'_1 &= b_1.
    \end{aligned}$$

2.  For $k=2,\ldots,n$, set $$\begin{aligned}
    d'_k &= d_k - \tfrac{l_k}{d'_{k-1}}u_{k-1}, & b'_k &= b_k - \tfrac{l_k}{d'_{k-1}}b'_{k-1}.
    \end{aligned}$$

3.  Set $$x_n = \tfrac{b'_n}{d'_n}.$$

4.  For $k=n-1,n-2,\ldots,1$ set
    $$x_k = \tfrac{1}{d'_k}(b'_k - u_k x_{k+1}).$$

## Implementing tridiagonal matrix solver in Python

-   Data structures
    -   NumPy array indexes start at 0. How do we deal with $u_1$ and
        $d_n$?
    -   Internal representation of sparse diagonal matrices a little
        complex for off-diagonals.
    -   Diagonals of sparse diagonal matrices not always stored in the
        same order.
-   Needs to be fast: vectorize where possible.
-   Other considerations?

## Implementation using NumPy arrays

-   We obtain the following procedure after:
    -   Starting at $k=0$ instead of $k=1$.
    -   Ignoring $l_0$ and $u_{n-1}$ (wasteful but not significant).

1.  Set $d'_0 = d_0$ and $b'_0 = b_0$.

2.  For $k=1,\ldots,n-1$, set $$\begin{aligned}
    d'_k &= d_k - \tfrac{l_k}{d'_{k-1}}u_{k-1}, & b'_k &= b_k - \tfrac{l_k}{d'_{k-1}}b'_{k-1}.
    \end{aligned}$$

3.  Set $x_{n-1} = b'_{n-1} / d'_{n-1}$.

4.  For $k=n-2,n-3,\ldots,0$ set
    $$x_k = \tfrac{1}{d'_k}(b'_k - u_k x_{k+1}).$$

------------------------------------------------------------------------

In [45]:
def tri_matrix_solver (u, d, l, b):
    """Solves tridiagonal system Ax = b where A is a tridiagonal matrix with diagonals u, d, l"""

    n = len(b)
    
    # new arrays for calculations - includes step 1
    ddash = np.array(d, dtype=float)
    bdash = np.array(b, dtype=float)

    # step 2
    for k in range(1,n):
        ddash[k] -= l[k]/ddash[k-1]*u[k-1]
        bdash[k] -= l[k]/ddash[k-1]*bdash[k-1]
        
    x = np.empty_like(b, dtype=float)

    #step 3
    x[-1] = bdash[-1]/ddash[-1]

    #step 4
    for k in range(n-2,-1,-1):
        x[k] = (bdash[k] - u[k] * x[k+1])/ddash[k]

    return x

u = np.tile(0.5,500)
d = np.tile(1, 500)
l = np.tile(0.5, 500)
b = np.ones(500)

x = tri_matrix_solver (u, d, l, b)
print("Size of x:", x.size)
print("First 5 elements of x:", x[:5])

Size of x: 500
First 5 elements of x: [0.99800399 0.00399202 0.99401198 0.00798403 0.99001996]

------------------------------------------------------------------------

## Implementation using SciPy sparse diagonal matrix

-   Create a $3\times 3$ matrix with 1 on the main diagonal, 0.5 on the
    diagonal above it and -0.5 on the diagonal below it:

In [46]:
u = np.tile(0.5, 2)
d = np.tile(1, 3)
l = np.tile(-0.5, 2)

G = ssparse.diags([u, d, l], offsets = [1, 0, -1])
print ("Diagonal offsets:", G.offsets)
print ("Data:\n", G.data)

Diagonal offsets: [ 1  0 -1]
Data:
 [[ 0.   0.5  0.5]
 [ 1.   1.   1. ]
 [-0.5 -0.5  0. ]]

In [47]:
H = ssparse.diags([l, d, u], offsets = [-1, 0, 1])
print ("Diagonal offsets:", H.offsets)
print ("Data:\n", H.data)

Diagonal offsets: [-1  0  1]
Data:
 [[-0.5 -0.5  0. ]
 [ 1.   1.   1. ]
 [ 0.   0.5  0.5]]

------------------------------------------------------------------------

### Notes

-   **Warning**: internal representation depends on how the sparse
    diagonal matrix is defined. We can’t assume anything about the order
    of diagonals.

-   Indices need to be adjusted to take internal representation into
    account:

    -   Contents of lower diagonal array shifted one step towards 0.
    -   Contents of upper diagonal array shifted one step away from 0.

------------------------------------------------------------------------

In [48]:
def tri_matrix_solver_sparse (A, b):
    """Solves tridiagonal system Ax = b. A is in sparse diagonal format with A.offsets == [1, 0, -1]"""

    if not (A.offsets == np.array([1,0,-1])).all():
        print("Diagonal offsets of A in incorrect order: should be 1,0,-1.")
        return None

    u = A.data[0]
    d = A.data[1]
    l = A.data[2]
    
    ddash = np.array(d, dtype=float)
    bdash = np.array(b, dtype=float)

    n = len(d)

    # step 2
    for k in range(1,n):
        ddash[k] -= l[k-1]/ddash[k-1]*u[k]
        bdash[k] -= l[k-1]/ddash[k-1]*bdash[k-1]
        
    x = np.empty_like(b, dtype=float)

    #step 3
    x[-1] = bdash[-1]/ddash[-1]

    #step 4
    for k in range(n-2,-1,-1):
        x[k] = (bdash[k] - u[k+1] * x[k+1])/ddash[k]

    return x

diagonals = [0.5, 1, 0.5]
offsets = [1, 0, -1]
E = ssparse.diags(diagonals, offsets = offsets, shape = (500,500))
x = tri_matrix_solver_sparse (E, b)
print("Size of x:", x.size)
print("First 5 elements of x:", x[:5])

Size of x: 500
First 5 elements of x: [0.99800399 0.00399202 0.99401198 0.00798403 0.99001996]

## Comparing efficiency for tridiagonal matrices

-   Let’s test the efficiency of the various methods for a large
    tridiagonal matrix.
-   Create a $500\times 500$ matrix with 1 on the main diagonal, 0.5 on
    the diagonal above it and 0.5 on the diagonal below it:

In [49]:
diagonals = [0.5, 1, 0.5]
offsets = [1, 0, -1]
E = ssparse.diags(diagonals, offsets = offsets, shape = (500,500))

-   The SciPy built-in methods require conversion to compressed sparse
    column format:

In [50]:
%timeit Ecsc = E.tocsc()

72 μs ± 2.79 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

-   Define 500-dimensional vector of 1s:

In [51]:
b = np.ones(500)

## NumPy built-in methods

-   First convert our matrix to a NumPy array:

In [52]:
%timeit Enp = E.toarray()

113 μs ± 2.21 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

-   Matrix inverse:

In [54]:
%timeit npla.inv(Enp) @ b

2.97 ms ± 52.4 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

-   Solving linear system:

In [55]:
%timeit npla.solve(Enp,b)

1.35 ms ± 74.4 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

-   This is faster than calculating the inverse of $E$ as a sparse
    matrix, but slower than all other methods (including tridiagonal
    solver).

## SciPy built-in methods

-   Matrix inverse:

In [56]:
%timeit ssla.inv(Ecsc) @ b

68.3 ms ± 2.49 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

-   Solving linear system:

In [57]:
%timeit ssla.spsolve(Ecsc, b)

123 μs ± 840 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

-   Creating a solver:

In [58]:
%timeit solve = ssla.factorized(Ecsc)

117 μs ± 2.56 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [60]:
%timeit solve(b)

9.88 μs ± 168 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)

-   `scipy.sparse.linalg.spsolve` is fastest when only one system needs
    to be solved. `scipy.sparse.linalg.factorized` is considerably
    faster if two or more systems need to be solved. `scipy.sparse.inv`
    is slow and to be avoided.

## Tridiagonal matrix solver

-   Tridiagonal matrix solver with NumPy arrays:

In [61]:
u = np.tile(0.5,500)
d = np.tile(1, 500)
l = np.tile(0.5, 500)

In [62]:
%timeit tri_matrix_solver (u, d, l, b)

676 μs ± 3.94 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

-   Tridiagonal matrix solver with diagonal sparse array

In [63]:
%timeit E = ssparse.diags(diagonals, offsets = offsets, shape = (500,500))

48.5 μs ± 1.39 μs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [64]:
%timeit tri_matrix_solver_sparse (E, b)

711 μs ± 8.28 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

-   Speed of two methods is comparable. The cost of creating the sparse
    matrix becomes insignificant if system needs to be solved many
    times.
-   Tridiagonal solver is faster than `scipy.sparse.inv` but slower than
    `scipy.sparse.linalg.spsolve` and `scipy.sparse.linalg.factorized`,
    which is recommended.

# References

NumPy Developers. 2024. “Linear Algebra (Numpy.linalg).”
<https://numpy.org/doc/stable/reference/routines.linalg.html#module-numpy.linalg>.

The SciPy Community. 2024a. “Scipy.sparse.csc_array.”
<https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csc_array.html#scipy.sparse.csc_array>.

———. 2024b. “Scipy.sparse.diags.”
<https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.diags.html#scipy.sparse.diags>.

———. 2024c. “Scipy.sparse.diags_array.”
<https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.diags_array.html#scipy.sparse.diags_array>.

———. 2024d. “Sparse Linear Algebra (Scipy.sparse.linalg).”
<https://docs.scipy.org/doc/scipy/reference/sparse.linalg.html#module-scipy.sparse.linalg>.

———. 2024e. “Sparse Matrices (Scipy.sparse).”
<https://docs.scipy.org/doc/scipy/reference/sparse.html>.